cli/lsp/language_server.rs

```diff
@@ -3284,9 +3313,10 @@ impl tower_lsp::LanguageServer for LanguageServer {
       {
         return;
       }
-      inner.diagnostics_server.state()
+      inner.diagnostics_state.clone()
     };
-    if !diagnostics_state.read().await.has_no_cache_diagnostic(uri) {
+    // Call this outside of the language server lock, it sync locks internally.
+    if !diagnostics_state.has_no_cache_diagnostics(uri) {
```

Comment:
This probably isn't necessary

Reply:
Okay for some reason when I remove it the `lsp_cache_on_save()` test freezes. Might have to do with the lock queueing?

Reply:
I'll take a look. I have to head out in 15 mins though.

Reply:
False alarm! I removed the `!` when I moved it and of course it freezes when reading diagnostics since none are published.

Reply:
Ah, ok good!